manpages/brew.1
.\" generated by kramdown
.TH "BREW" "1" "July 2024" "Homebrew"
.SH NAME
brew \- The Missing Package Manager for macOS (or Linux)
.SH "SYNOPSIS"
\fBbrew\fP \fB\-\-version\fP
.br
\fBbrew\fP \fIcommand\fP [\fB\-\-verbose\fP|\fB\-v\fP] [\fIoptions\fP] [\fIformula\fP] \.\.\.
.SH "DESCRIPTION"
Homebrew is the easiest and most flexible way to install the UNIX tools Apple didn\[u2019]t include with macOS\. It can also install software not packaged for your Linux distribution without requiring \fBsudo\fP\&\.
.SH "TERMINOLOGY"
.TP
\fBformula\fP
Homebrew package definition that builds from upstream sources
.TP
\fBcask\fP
Homebrew package definition that installs macOS native applications
.TP
\fBprefix\fP
path in which Homebrew is installed, e\.g\. \fB/usr/local\fP
.TP
\fBkeg\fP
installation destination directory of a given \fBformula\fP version, e\.g\. \fB/usr/local/Cellar/foo/0\.1\fP
.TP
\fBrack\fP
directory containing one or more versioned \fBkegs\fP, e\.g\. \fB/usr/local/Cellar/foo\fP
.TP
\fBkeg\-only\fP
a \fBformula\fP is \fIkeg\-only\fP if it is not symlinked into Homebrew\[u2019]s prefix
.TP
\fBopt prefix\fP
a symlink to the active version of a \fBkeg\fP, e\.g\. \fB/usr/local/opt/foo\fP
.TP
\fBCellar\fP
directory containing one or more named \fBracks\fP, e\.g\. \fB/usr/local/Cellar\fP
.TP
\fBCaskroom\fP
directory containing one or more named \fBcasks\fP, e\.g\. \fB/usr/local/Caskroom\fP
.TP
\fBexternal command\fP
\fBbrew\fP subcommand defined outside of the Homebrew/brew GitHub repository
.TP
\fBtap\fP
directory (and usually Git repository) of \fBformulae\fP, \fBcasks\fP and/or \fBexternal commands\fP
.TP
\fBbottle\fP
pre\-built \fBkeg\fP poured into a \fBrack\fP of the \fBCellar\fP instead of building from upstream sources
.SH "ESSENTIAL COMMANDS"
For the full command list, see the \fICOMMANDS\fP section\.
.P
With \fB\-\-verbose\fP or \fB\-\-debug\fP, many commands print extra debugging information\. Note that these options should only appear after a command\.
.P
Some command behaviour can be customised with environment variables; see the \fIENVIRONMENT\fP section\.
.SS "\fBinstall\fP \fIformula\fP"
Install \fIformula\fP\&\.
.P
\fIformula\fP is usually the name of the formula to install, but it has other syntaxes which are listed in the \fISPECIFYING FORMULAE\fP section\.
.SS "\fBuninstall\fP \fIformula\fP"
Uninstall \fIformula\fP\&\.
.SS "\fBlist\fP"
List all installed formulae\.
.SS "\fBsearch\fP \fR[\fItext\fP|\fB/\fP\fItext\fP\fB/\fP]"
Perform a substring search of cask tokens and formula names for \fItext\fP\&\. If \fItext\fP is flanked by slashes, it is interpreted as a regular expression\. The search for \fItext\fP is extended online to \fBhomebrew/core\fP and \fBhomebrew/cask\fP\&\. If no search term is provided, all locally available formulae are listed\.
.SH "COMMANDS"
.SS "\fBanalytics\fP \fR[\fIsubcommand\fP]"
Control Homebrew\[u2019]s anonymous aggregate user behaviour analytics\. Read more at 
.UR https://docs\.brew\.sh/Analytics
.UE \&\.
.TP
\fBbrew analytics\fP [\fBstate\fP]
Display the current state of Homebrew\[u2019]s analytics\.
.TP
\fBbrew analytics\fP (\fBon\fP|\fBoff\fP)
Turn Homebrew\[u2019]s analytics on or off respectively\.
.SS "\fBautoremove\fP \fR[\fB\-\-dry\-run\fP]"
Uninstall formulae that were only installed as a dependency of another formula and are now no longer needed\.
.TP
\fB\-n\fP, \fB\-\-dry\-run\fP
List what would be uninstalled, but do not actually uninstall anything\.
.SS "\fBcasks\fP"
List all locally installable casks including short names\.
.SS "\fBcleanup\fP \fR[\fIoptions\fP] \fR[\fIformula\fP|\fIcask\fP \.\.\.]"
Remove stale lock files and outdated downloads for all formulae and casks, and remove old versions of installed formulae\. If arguments are specified, only do this for the given formulae and casks\. Removes all downloads more than 120 days old\. This can be adjusted with \fBHOMEBREW_CLEANUP_MAX_AGE_DAYS\fP\&\.
.TP
\fB\-\-prune\fP
Remove all cache files older than specified \fIdays\fP\&\. If you want to remove everything, use \fB\-\-prune=all\fP\&\.
.TP
\fB\-n\fP, \fB\-\-dry\-run\fP
Show what would be removed, but do not actually remove anything\.
.TP
\fB\-s\fP, \fB\-\-scrub\fP
Scrub the cache, including downloads for even the latest versions\. Note that downloads for any installed formulae or casks will still not be deleted\. If you want to delete those too: \fBrm \-rf "$(brew \-\-cache)"\fP
.TP
\fB\-\-prune\-prefix\fP
Only prune the symlinks and directories from the prefix and remove no other files\.
.SS "\fBcommand\fP \fIcommand\fP \fR[\.\.\.]"
Display the path to the file being used when invoking \fBbrew\fP \fIcmd\fP\&\.
.SS "\fBcommands\fP \fR[\fB\-\-quiet\fP] \fR[\fB\-\-include\-aliases\fP]"
Show lists of built\-in and external commands\.
.TP
\fB\-q\fP, \fB\-\-quiet\fP
List only the names of commands without category headers\.
.TP
\fB\-\-include\-aliases\fP
Include aliases of internal commands\.
.SS "\fBcompletions\fP \fR[\fIsubcommand\fP]"
Control whether Homebrew automatically links external tap shell completion files\. Read more at 
.UR https://docs\.brew\.sh/Shell\-Completion
.UE \&\.
.TP
\fBbrew completions\fP [\fBstate\fP]
Display the current state of Homebrew\[u2019]s completions\.
.TP
\fBbrew completions\fP (\fBlink\fP|\fBunlink\fP)
Link or unlink Homebrew\[u2019]s completions\.
.SS "\fBconfig\fP, \fB\-\-config\fP"
Show Homebrew and system configuration info useful for debugging\. If you file a bug report, you will be required to provide this information\.
.SS "\fBdeps\fP \fR[\fIoptions\fP] \fR[\fIformula\fP|\fIcask\fP \.\.\.]"
Show dependencies for \fIformula\fP\&\. When given multiple formula arguments, show the intersection of dependencies for each formula\. By default, \fBdeps\fP shows all required and recommended dependencies\.
.P
If any version of each formula argument is installed and no other options are passed, this command displays their actual runtime dependencies (similar to \fBbrew linkage\fP), which may differ from the current versions\[u2019] stated dependencies if the installed versions are outdated\.
.P
\fINote:\fP \fB\-\-missing\fP and \fB\-\-skip\-recommended\fP have precedence over \fB\-\-include\-*\fP\&\.
.TP
\fB\-n\fP, \fB\-\-topological\fP
Sort dependencies in topological order\.
.TP
\fB\-1\fP, \fB\-\-direct\fP
Show only the direct dependencies declared in the formula\.
.TP
\fB\-\-union\fP
Show the union of dependencies for multiple \fIformula\fP, instead of the intersection\.
.TP
\fB\-\-full\-name\fP
List dependencies by their full name\.
.TP
\fB\-\-include\-build\fP
Include \fB:build\fP dependencies for \fIformula\fP\&\.
.TP
\fB\-\-include\-optional\fP
Include \fB:optional\fP dependencies for \fIformula\fP\&\.
.TP
\fB\-\-include\-test\fP
Include \fB:test\fP dependencies for \fIformula\fP (non\-recursive)\.
.TP
\fB\-\-skip\-recommended\fP
Skip \fB:recommended\fP dependencies for \fIformula\fP\&\.
.TP
\fB\-\-include\-requirements\fP
Include requirements in addition to dependencies for \fIformula\fP\&\.
.TP
\fB\-\-tree\fP
Show dependencies as a tree\. When given multiple formula arguments, show individual trees for each formula\.
.TP
\fB\-\-graph\fP
Show dependencies as a directed graph\.
.TP
\fB\-\-dot\fP
Show text\-based graph description in DOT format\.
.TP
\fB\-\-annotate\fP
Mark any build, test, implicit, optional, or recommended dependencies as such in the output\.
.TP
\fB\-\-installed\fP
List dependencies for formulae that are currently installed\. If \fIformula\fP is specified, list only its dependencies that are currently installed\.
.TP
\fB\-\-missing\fP
Show only missing dependencies\.
.TP
\fB\-\-eval\-all\fP
Evaluate all available formulae and casks, whether installed or not, to list their dependencies\.
.TP
\fB\-\-for\-each\fP
Switch into the mode used by the \fB\-\-eval\-all\fP option, but only list dependencies for each provided \fIformula\fP, one formula per line\. This is used for debugging the \fB\-\-installed\fP/\fB\-\-eval\-all\fP display mode\.
.TP
\fB\-\-HEAD\fP
Show dependencies for HEAD version instead of stable version\.
.TP
\fB\-\-os\fP
Show dependencies for the given operating system\.
.TP
\fB\-\-arch\fP
Show dependencies for the given CPU architecture\.
.TP
\fB\-\-formula\fP
Treat all named arguments as formulae\.
.TP
\fB\-\-cask\fP
Treat all named arguments as casks\.
.SS "\fBdesc\fP \fR[\fIoptions\fP] \fIformula\fP|\fIcask\fP|\fItext\fP|\fB/\fP\fIregex\fP\fB/\fP \fR[\.\.\.]"
Display \fIformula\fP\[u2019]s name and one\-line description\. The cache is created on the first search, making that search slower than subsequent ones\.
.TP
\fB\-s\fP, \fB\-\-search\fP
Search both names and descriptions for \fItext\fP\&\. If \fItext\fP is flanked by slashes, it is interpreted as a regular expression\.
.TP
\fB\-n\fP, \fB\-\-name\fP
Search just names for \fItext\fP\&\. If \fItext\fP is flanked by slashes, it is interpreted as a regular expression\.
.TP
\fB\-d\fP, \fB\-\-description\fP
Search just descriptions for \fItext\fP\&\. If \fItext\fP is flanked by slashes, it is interpreted as a regular expression\.
.TP
\fB\-\-eval\-all\fP
Evaluate all available formulae and casks, whether installed or not, to search their descriptions\. Implied if \fBHOMEBREW_EVAL_ALL\fP is set\.
.TP
\fB\-\-formula\fP
Treat all named arguments as formulae\.
.TP
\fB\-\-cask\fP
Treat all named arguments as casks\.
.SS "\fBdeveloper\fP \fR[\fIsubcommand\fP]"
Control Homebrew\[u2019]s developer mode\. When developer mode is enabled, \fBbrew update\fP will update Homebrew to the latest commit on the \fBmaster\fP branch instead of the latest stable version along with some other behaviour changes\.
.TP
\fBbrew developer\fP [\fBstate\fP]
Display the current state of Homebrew\[u2019]s developer mode\.
.TP
\fBbrew developer\fP (\fBon\fP|\fBoff\fP)
Turn Homebrew\[u2019]s developer mode on or off respectively\.
.SS "\fBdocs\fP"
Open Homebrew\[u2019]s online documentation at 
.UR https://docs\.brew\.sh
.UE
in a browser\.
.SS "\fBdoctor\fP, \fBdr\fP \fR[\fB\-\-list\-checks\fP] \fR[\fB\-\-audit\-debug\fP] \fR[\fIdiagnostic_check\fP \.\.\.]"
Check your system for potential problems\. Will exit with a non\-zero status if any potential problems are found\.
.P
Please note that these warnings are just used to help the Homebrew maintainers with debugging if you file an issue\. If everything you use Homebrew for is working fine: please don\[u2019]t worry or file an issue; just ignore this\.
.TP
\fB\-\-list\-checks\fP
List all audit methods, which can be run individually if provided as arguments\.
.TP
\fB\-D\fP, \fB\-\-audit\-debug\fP
Enable debugging and profiling of audit methods\.
.SS "\fBfetch\fP \fR[\fIoptions\fP] \fIformula\fP|\fIcask\fP \fR[\.\.\.]"
Download a bottle (if available) or source packages for \fIformula\fPe and binaries for \fIcask\fPs\. For files, also print SHA\-256 checksums\.
.TP
\fB\-\-os\fP
Download for the given operating system\. (Pass \fBall\fP to download for all operating systems\.)
.TP
\fB\-\-arch\fP
Download for the given CPU architecture\. (Pass \fBall\fP to download for all architectures\.)
.TP
\fB\-\-bottle\-tag\fP
Download a bottle for given tag\.
.TP
\fB\-\-HEAD\fP
Fetch HEAD version instead of stable version\.
.TP
\fB\-f\fP, \fB\-\-force\fP
Remove a previously cached version and re\-fetch\.
.TP
\fB\-v\fP, \fB\-\-verbose\fP
Do a verbose VCS checkout, if the URL represents a VCS\. This is useful for seeing if an existing VCS cache has been updated\.
.TP
\fB\-\-retry\fP
Retry if downloading fails or re\-download if the checksum of a previously cached version no longer matches\. Tries at most 5 times with exponential backoff\.
.TP
\fB\-\-deps\fP
Also download dependencies for any listed \fIformula\fP\&\.
.TP
\fB\-s\fP, \fB\-\-build\-from\-source\fP
Download source packages rather than a bottle\.
.TP
\fB\-\-build\-bottle\fP
Download source packages (for eventual bottling) rather than a bottle\.
.TP
\fB\-\-force\-bottle\fP
Download a bottle if it exists for the current or newest version of macOS, even if it would not be used during installation\.
.TP
\fB\-\-[no\-]quarantine\fP
Disable/enable quarantining of downloads (default: enabled)\.
.TP
\fB\-\-formula\fP
Treat all named arguments as formulae\.
.TP
\fB\-\-cask\fP
Treat all named arguments as casks\.
.SS "\fBformulae\fP"
List all locally installable formulae including short names\.
.SS "\fBgist\-logs\fP \fR[\fIoptions\fP] \fIformula\fP"
Upload logs for a failed build of \fIformula\fP to a new Gist\. Presents an error message if no logs are found\.
.TP
\fB\-\-with\-hostname\fP
Include the hostname in the Gist\.
.TP
\fB\-n\fP, \fB\-\-new\-issue\fP
Automatically create a new issue in the appropriate GitHub repository after creating the Gist\.
.TP
\fB\-p\fP, \fB\-\-private\fP
The Gist will be marked private and will not appear in listings but will be accessible with its link\.
.SS "\fBhelp\fP \fR[\fIcommand\fP \.\.\.]"
Outputs the usage instructions for \fBbrew\fP \fIcommand\fP\&\. Equivalent to \fBbrew \-\-help\fP \fIcommand\fP\&\.
.SS "\fBhome\fP, \fBhomepage\fP \fR[\fB\-\-formula\fP] \fR[\fB\-\-cask\fP] \fR[\fIformula\fP|\fIcask\fP \.\.\.]"
Open a \fIformula\fP or \fIcask\fP\[u2019]s homepage in a browser, or open Homebrew\[u2019]s own homepage if no argument is provided\.
.TP
\fB\-\-formula\fP
Treat all named arguments as formulae\.
.TP
\fB\-\-cask\fP
Treat all named arguments as casks\.
.SS "\fBinfo\fP, \fBabv\fP \fR[\fIoptions\fP] \fR[\fIformula\fP|\fIcask\fP \.\.\.]"
Display brief statistics for your Homebrew installation\. If a \fIformula\fP or \fIcask\fP is provided, show summary of information about it\.
.TP
\fB\-\-analytics\fP
List global Homebrew analytics data or, if specified, installation and build error data for \fIformula\fP (provided neither \fBHOMEBREW_NO_ANALYTICS\fP nor \fBHOMEBREW_NO_GITHUB_API\fP are set)\.
.TP
\fB\-\-days\fP
How many days of analytics data to retrieve\. The value for \fIdays\fP must be \fB30\fP, \fB90\fP or \fB365\fP\&\. The default is \fB30\fP\&\.
.TP
\fB\-\-category\fP
Which type of analytics data to retrieve\. The value for \fIcategory\fP must be \fBinstall\fP, \fBinstall\-on\-request\fP or \fBbuild\-error\fP; \fBcask\-install\fP or \fBos\-version\fP may be specified if \fIformula\fP is not\. The default is \fBinstall\fP\&\.
.TP
\fB\-\-github\fP
Open the GitHub source page for \fIformula\fP and \fIcask\fP in a browser\. To view the history locally: \fBbrew log \-p\fP \fIformula\fP or \fIcask\fP
.TP
\fB\-\-json\fP
Print a JSON representation\. Currently the default value for \fIversion\fP is \fBv1\fP for \fIformula\fP\&\. For \fIformula\fP and \fIcask\fP use \fBv2\fP\&\. See the docs for examples of using the JSON output: 
.UR https://docs\.brew\.sh/Querying\-Brew
.UE
.TP
\fB\-\-installed\fP
Print JSON of formulae that are currently installed\.
.TP
\fB\-\-eval\-all\fP
Evaluate all available formulae and casks, whether installed or not, to print their JSON\. Implied if \fBHOMEBREW_EVAL_ALL\fP is set\.
.TP
\fB\-\-variations\fP
Include the variations hash in each formula\[u2019]s JSON output\.
.TP
\fB\-v\fP, \fB\-\-verbose\fP
Show more verbose analytics data for \fIformula\fP\&\.
.TP
\fB\-\-formula\fP
Treat all named arguments as formulae\.
.TP
\fB\-\-cask\fP
Treat all named arguments as casks\.
.SS "\fBinstall\fP \fR[\fIoptions\fP] \fIformula\fP|\fIcask\fP \fR[\.\.\.]"
Install a \fIformula\fP or \fIcask\fP\&\. Additional options specific to a \fIformula\fP may be appended to the command\.
.P
Unless \fBHOMEBREW_NO_INSTALLED_DEPENDENTS_CHECK\fP is set, \fBbrew upgrade\fP or \fBbrew reinstall\fP will be run for outdated dependents and dependents with broken linkage, respectively\.
.P
Unless \fBHOMEBREW_NO_INSTALL_CLEANUP\fP is set, \fBbrew cleanup\fP will then be run for the installed formulae or, every 30 days, for all formulae\.
.P
Unless \fBHOMEBREW_NO_INSTALL_UPGRADE\fP is set, \fBbrew install\fP \fIformula\fP will upgrade \fIformula\fP if it is already installed but outdated\.
.TP
\fB\-d\fP, \fB\-\-debug\fP
If brewing fails, open an interactive debugging session with access to IRB or a shell inside the temporary build directory\.
.TP
\fB\-\-display\-times\fP
Print install times for each package at the end of the run\.
.TP
\fB\-f\fP, \fB\-\-force\fP
Install formulae without checking for previously installed keg\-only or non\-migrated versions\. When installing casks, overwrite existing files (binaries and symlinks are excluded, unless originally from the same cask)\.
.TP
\fB\-v\fP, \fB\-\-verbose\fP
Print the verification and post\-install steps\.
.TP
\fB\-n\fP, \fB\-\-dry\-run\fP
Show what would be installed, but do not actually install anything\.
.TP
\fB\-\-formula\fP
Treat all named arguments as formulae\.
.TP
\fB\-\-ignore\-dependencies\fP
An unsupported Homebrew development option to skip installing any dependencies of any kind\. If the dependencies are not already present, the formula will have issues\. If you\[u2019]re not developing Homebrew, consider adjusting your PATH rather than using this option\.
.TP
\fB\-\-only\-dependencies\fP
Install the dependencies with specified options but do not install the formula itself\.
.TP
\fB\-\-cc\fP
Attempt to compile using the specified \fIcompiler\fP, which should be the name of the compiler\[u2019]s executable, e\.g\. \fBgcc\-7\fP for GCC 7\. In order to use LLVM\[u2019]s clang, specify \fBllvm_clang\fP\&\. To use the Apple\-provided clang, specify \fBclang\fP\&\. This option will only accept compilers that are provided by Homebrew or bundled with macOS\. Please do not file issues if you encounter errors while using this option\.
.TP
\fB\-s\fP, \fB\-\-build\-from\-source\fP
Compile \fIformula\fP from source even if a bottle is provided\. Dependencies will still be installed from bottles if they are available\.
.TP
\fB\-\-force\-bottle\fP
Install from a bottle if it exists for the current or newest version of macOS, even if it would not normally be used for installation\.
.TP
\fB\-\-include\-test\fP
Install testing dependencies required to run \fBbrew test\fP \fIformula\fP\&\.
.TP
\fB\-\-HEAD\fP
If \fIformula\fP defines it, install the HEAD version, aka\. main, trunk, unstable, master\.
.TP
\fB\-\-fetch\-HEAD\fP
Fetch the upstream repository to detect if the HEAD installation of the formula is outdated\. Otherwise, the repository\[u2019]s HEAD will only be checked for updates when a new stable or development version has been released\.
.TP
\fB\-\-keep\-tmp\fP
Retain the temporary files created during installation\.
.TP
\fB\-\-debug\-symbols\fP
Generate debug symbols on build\. Source will be retained in a cache directory\.
.TP
\fB\-\-build\-bottle\fP
Prepare the formula for eventual bottling during installation, skipping any post\-install steps\.
.TP
\fB\-\-skip\-post\-install\fP
Install but skip any post\-install steps\.
.TP
\fB\-\-bottle\-arch\fP
Optimise bottles for the specified architecture rather than the oldest architecture supported by the version of macOS the bottles are built on\.
.TP
\fB\-i\fP, \fB\-\-interactive\fP
Download and patch \fIformula\fP, then open a shell\. This allows the user to run \fB\&\./configure \-\-help\fP and otherwise determine how to turn the software package into a Homebrew package\.
.TP
\fB\-g\fP, \fB\-\-git\fP
Create a Git repository, useful for creating patches to the software\.
.TP
\fB\-\-overwrite\fP
Delete files that already exist in the prefix while linking\.
.TP
\fB\-\-cask\fP
Treat all named arguments as casks\.
.TP
\fB\-\-[no\-]binaries\fP
Disable/enable linking of helper executables (default: enabled)\.
.TP
\fB\-\-require\-sha\fP
Require all casks to have a checksum\.
.TP
\fB\-\-[no\-]quarantine\fP
Disable/enable quarantining of downloads (default: enabled)\.
.TP
\fB\-\-adopt\fP
Adopt existing artifacts in the destination that are identical to those being installed\. Cannot be combined with \fB\-\-force\fP\&\.
.TP
\fB\-\-skip\-cask\-deps\fP
Skip installing cask dependencies\.
.TP
\fB\-\-zap\fP
For use with \fBbrew reinstall \-\-cask\fP\&\. Remove all files associated with a cask\. \fIMay remove files which are shared between applications\.\fP
.SS "\fBleaves\fP \fR[\fB\-\-installed\-on\-request\fP] \fR[\fB\-\-installed\-as\-dependency\fP]"

docs/Releases.md
# Releases

Since Homebrew 1.0.0 most Homebrew users (those who haven't run a `dev-cmd` or set `HOMEBREW_DEVELOPER=1` which is ~99.9% based on analytics data) require tags on the [Homebrew/brew repository](https://github.com/homebrew/brew) in order to receive new versions of Homebrew. There are a few steps in making a new Homebrew release:

1. Check if there is anything pressing that needs to be fixed or merged before the next release in:
   - [`Homebrew/brew` pull requests](https://github.com/homebrew/brew/pulls)
   - [`Homebrew/brew` issues](https://github.com/homebrew/brew/issues)
   - [`Homebrew/homebrew-core` issues](https://github.com/homebrew/homebrew-core/issues)
   - [Homebrew/discussions (forum)](https://github.com/orgs/Homebrew/discussions)

    If so, fix and merge these changes.

2. Ensure that:
   - no code changes have happened for at least a couple of hours (ideally 4 hours),
   - at least one Homebrew/homebrew-core pull request CI job has completed successfully,
   - the state of the Homebrew/brew `master` CI job is clear (i.e. main jobs green or green after rerunning)
   - you are confident there are no major regressions on the current `master` branch.

3. Run `brew release` to create a new draft release. For major or minor version bumps, pass `--major` or `--minor`, respectively.

4. Publish the draft release on [GitHub](https://github.com/Homebrew/brew/releases).

If this is a major or minor release (e.g. X.0.0 or X.Y.0) then there are a few more steps:

1. Before creating the tag you should:
   - delete any `odisabled` code,
   - make any `odeprecated` code `odisabled`,
   - uncomment any `# odeprecated` code
   - add any new `odeprecations` that are desired.

   Also delete any command argument definitions that pass `replacement: ...`.

2. Write up a release notes blog post for <https://brew.sh> (e.g. [brew.sh#319](https://github.com/Homebrew/brew.sh/pull/319)). This should use the output from `brew release [--major|--minor]` as input but have the wording adjusted to be more human readable and explain not just what has changed but why.

3. When the release has shipped and the blog post has been merged, tweet the blog post as the [@MacHomebrew Twitter account](https://twitter.com/MacHomebrew) or tweet it yourself and retweet it with the @MacHomebrew Twitter account (credentials are in 1Password).

4. Consider whether to submit it to other sources, e.g. Hacker News, Reddit.
   - Pros: gets a wider reach and user feedback
   - Cons: negative comments are common and people take this as a chance to complain about Homebrew (regardless of their usage)

Please do not manually create a release based on older commits on the `master` branch. It's very hard to judge whether these have been sufficiently tested by users or if they will cause negative side effects with the current state of Homebrew/homebrew-core. If a new branch is needed ASAP but there are things on `master` that cannot be released yet (e.g. new deprecations and you want to make a patch release) then revert the relevant PRs, follow the process above and then revert the reverted PRs to reapply them on `master`.


Library/Homebrew/brew.rb
# typed: strict
# frozen_string_literal: true

# `HOMEBREW_STACKPROF` should be set via `brew prof --stackprof`, not manually.
if ENV["HOMEBREW_STACKPROF"]
  require "rubygems"
  require "stackprof"
  StackProf.start(mode: :wall, raw: true)
end

raise "HOMEBREW_BREW_FILE was not exported! Please call bin/brew directly!" unless ENV["HOMEBREW_BREW_FILE"]
if $PROGRAM_NAME != __FILE__ && !$PROGRAM_NAME.end_with?("/bin/ruby-prof")
  raise "#{__FILE__} must not be loaded via `require`."
end

std_trap = trap("INT") { exit! 130 } # no backtrace thanks

require_relative "global"

begin
  trap("INT", std_trap) # restore default CTRL-C handler

  if ENV["CI"]
    $stdout.sync = true
    $stderr.sync = true
  end

  empty_argv = ARGV.empty?
  help_flag_list = %w[-h --help --usage -?]
  help_flag = !ENV["HOMEBREW_HELP"].nil?
  help_cmd_index = T.let(nil, T.nilable(Integer))
  cmd = T.let(nil, T.nilable(String))

  ARGV.each_with_index do |arg, i|
    break if help_flag && cmd

    if arg == "help" && !cmd
      # Command-style help: `help <cmd>` is fine, but `<cmd> help` is not.
      help_flag = true
      help_cmd_index = i
    elsif !cmd && help_flag_list.exclude?(arg)
      cmd = ARGV.delete_at(i)
      cmd = Commands::HOMEBREW_INTERNAL_COMMAND_ALIASES.fetch(cmd, cmd)
    end
  end

  ARGV.delete_at(help_cmd_index) if help_cmd_index

  require "cli/parser"
  args = Homebrew::CLI::Parser.new.parse(ARGV.dup.freeze, ignore_invalid_options: true)
  Context.current = args.context

  path = PATH.new(ENV.fetch("PATH"))
  homebrew_path = PATH.new(ENV.fetch("HOMEBREW_PATH"))

  # Add shared wrappers.
  path.prepend(HOMEBREW_SHIMS_PATH/"shared")
  homebrew_path.prepend(HOMEBREW_SHIMS_PATH/"shared")

  ENV["PATH"] = path.to_s

  require "abstract_command"
  require "commands"
  require "settings"

  internal_cmd = Commands.valid_internal_cmd?(cmd) || Commands.valid_internal_dev_cmd?(cmd) if cmd

  unless internal_cmd
    # Add contributed commands to PATH before checking.
    homebrew_path.append(Tap.cmd_directories)

    # External commands expect a normal PATH
    ENV["PATH"] = homebrew_path.to_s
  end

  # Usage instructions should be displayed if and only if one of:
  # - a help flag is passed AND a command is matched
  # - a help flag is passed AND there is no command specified
  # - no arguments are passed
  if empty_argv || help_flag
    require "help"
    Homebrew::Help.help cmd, remaining_args: args.remaining, empty_argv:
    # `Homebrew::Help.help` never returns, except for unknown commands.
  end

  if internal_cmd || Commands.external_ruby_v2_cmd_path(cmd)
    cmd = T.must(cmd)
    cmd_class = Homebrew::AbstractCommand.command(cmd)
    if cmd_class
      command_instance = cmd_class.new
      Utils::Analytics.report_command_run(command_instance)
      command_instance.run
    else
      Homebrew.public_send Commands.method_name(cmd)
    end
  elsif (path = Commands.external_ruby_cmd_path(cmd))
    require?(path)
    exit Homebrew.failed? ? 1 : 0
  elsif Commands.external_cmd_path(cmd)
    %w[CACHE LIBRARY_PATH].each do |env|
      ENV["HOMEBREW_#{env}"] = Object.const_get(:"HOMEBREW_#{env}").to_s
    end
    exec "brew-#{cmd}", *ARGV
  else
    possible_tap = OFFICIAL_CMD_TAPS.find { |_, cmds| cmds.include?(cmd) }
    possible_tap = Tap.fetch(possible_tap.first) if possible_tap

    if !possible_tap ||
       possible_tap.installed? ||
       (blocked_tap = Tap.untapped_official_taps.include?(possible_tap.name))
      if blocked_tap
        onoe <<~EOS
          `brew #{cmd}` is unavailable because #{possible_tap.name} was manually untapped.
          Run `brew tap #{possible_tap.name}` to reenable `brew #{cmd}`.
        EOS
      end
      # Check for cask explicitly because it's very common in old guides
      odie "`brew cask` is no longer a `brew` command. Use `brew <command> --cask` instead." if cmd == "cask"
      odie "Unknown command: brew #{cmd}"
    end

    # Unset HOMEBREW_HELP to avoid confusing the tap
    with_env HOMEBREW_HELP: nil do
      tap_commands = []
      if (File.exist?("/.dockerenv") ||
         Homebrew.running_as_root? ||
         ((cgroup = Utils.popen_read("cat", "/proc/1/cgroup").presence) &&
          %w[azpl_job actions_job docker garden kubepods].none? { |type| cgroup.include?(type) })) &&
         Homebrew.running_as_root_but_not_owned_by_root?
        tap_commands += %W[/usr/bin/sudo -u ##{Homebrew.owner_uid}]
      end
      quiet_arg = args.quiet? ? "--quiet" : nil
      tap_commands += [HOMEBREW_BREW_FILE, "tap", *quiet_arg, possible_tap.name]
      safe_system(*tap_commands)
    end

    ARGV << "--help" if help_flag
    exec HOMEBREW_BREW_FILE, cmd, *ARGV
  end
rescue UsageError => e
  require "help"
  Homebrew::Help.help cmd, remaining_args: args&.remaining, usage_error: e.message
rescue SystemExit => e
  onoe "Kernel.exit" if args&.debug? && !e.success?
  $stderr.puts Utils::Backtrace.clean(e) if args&.debug? || ARGV.include?("--debug")
  raise
rescue Interrupt
  $stderr.puts # seemingly a newline is typical
  exit 130
rescue BuildError => e
  Utils::Analytics.report_build_error(e)
  e.dump(verbose: args&.verbose? || false)

  if OS.unsupported_configuration?
    $stderr.puts "#{Tty.bold}Do not report this issue: you are running in an unsupported configuration.#{Tty.reset}"
  elsif e.formula.head? || e.formula.deprecated? || e.formula.disabled?
    reason = if e.formula.head?
      "was built from an unstable upstream --HEAD"
    elsif e.formula.deprecated?
      "is deprecated"
    elsif e.formula.disabled?
      "is disabled"
    end
    $stderr.puts <<~EOS
      #{e.formula.name}'s formula #{reason}.
      This build failure is expected behaviour.
      Do not create issues about this on Homebrew's GitHub repositories.
      Any opened issues will be immediately closed without response.
      Do not ask for help from Homebrew or its maintainers on social media.
      You may ask for help in Homebrew's discussions but are unlikely to receive a response.
      Try to figure out the problem yourself and submit a fix as a pull request.
      We will review it but may or may not accept it.
    EOS

  end

  exit 1
rescue RuntimeError, SystemCallError => e
  raise if e.message.empty?

  onoe e
  $stderr.puts Utils::Backtrace.clean(e) if args&.debug? || ARGV.include?("--debug")

  exit 1
rescue Exception => e # rubocop:disable Lint/RescueException
  onoe e

  method_deprecated_error = e.is_a?(MethodDeprecatedError)
  $stderr.puts Utils::Backtrace.clean(e) if args&.debug? || ARGV.include?("--debug") || !method_deprecated_error

  if OS.unsupported_configuration?
    $stderr.puts "#{Tty.bold}Do not report this issue: you are running in an unsupported configuration.#{Tty.reset}"
  elsif Homebrew::EnvConfig.no_auto_update? &&
        (fetch_head = HOMEBREW_REPOSITORY/".git/FETCH_HEAD") &&
        (!fetch_head.exist? || (fetch_head.mtime.to_date < Date.today))
    $stderr.puts "#{Tty.bold}You have disabled automatic updates and have not updated today.#{Tty.reset}"
    $stderr.puts "#{Tty.bold}Do not report this issue until you've run `brew update` and tried again.#{Tty.reset}"
  elsif (issues_url = (method_deprecated_error && e.issues_url) || Utils::Backtrace.tap_error_url(e))
    $stderr.puts "If reporting this issue please do so at (not Homebrew/brew or Homebrew/homebrew-core):"
    $stderr.puts "  #{Formatter.url(issues_url)}"
  elsif internal_cmd
    $stderr.puts "#{Tty.bold}Please report this issue:#{Tty.reset}"
    $stderr.puts "  #{Formatter.url(OS::ISSUES_URL)}"
  end

  exit 1
else
  exit 1 if Homebrew.failed?
ensure
  if ENV["HOMEBREW_STACKPROF"]
    StackProf.stop
    StackProf.results("prof/stackprof.dump")
  end
end
